widget/button.go

```diff
@@ -29,45 +29,152 @@ func (b *buttonRenderer) padding() fyne.Size {
 // MinSize calculates the minimum size of a button.
 // This is based on the contained text, any icon that is set and a standard
 // amount of padding added.
-func (b *buttonRenderer) MinSize() fyne.Size {
+func (b *buttonRenderer) MinSize() (size fyne.Size) {
+	hasIcon := b.icon != nil
+	hasLabel := b.label.Text != ""
+	iconSize := fyne.NewSize(theme.IconInlineSize(), theme.IconInlineSize())
 	labelSize := b.label.MinSize()
-	contentHeight := fyne.Max(labelSize.Height, theme.IconInlineSize())
-	contentWidth := 0
-	if b.icon != nil {
-		contentWidth += theme.IconInlineSize()
-	}
-	if b.button.Text != "" {
-		if b.icon != nil {
-			contentWidth += theme.Padding()
+	if hasLabel {
+		size.Width = labelSize.Width
+	}
+	if hasIcon {
+		if hasLabel {
+			size.Width += theme.Padding()
 		}
-		contentWidth += labelSize.Width
+		size.Width += iconSize.Width
 	}
-	return fyne.NewSize(contentWidth, contentHeight).Add(b.padding())
+	size.Height = fyne.Max(labelSize.Height, iconSize.Height)
+	size = size.Add(b.padding())
+	return
 }
 
 // Layout the components of the button widget
 func (b *buttonRenderer) Layout(size fyne.Size) {
 	b.LayoutShadow(size, fyne.NewPos(0, 0))
-	if b.button.Text != "" {
-		padding := b.padding()
-		innerSize := size.Subtract(padding)
-		innerOffset := fyne.NewPos(padding.Width/2, padding.Height/2)
-
-		labelSize := b.label.MinSize()
-		contentWidth := labelSize.Width
-
-		if b.icon != nil {
-			contentWidth += theme.Padding() + theme.IconInlineSize()
-			iconOffset := fyne.NewPos((innerSize.Width-contentWidth)/2, (innerSize.Height-theme.IconInlineSize())/2)
-			b.icon.Resize(fyne.NewSize(theme.IconInlineSize(), theme.IconInlineSize()))
-			b.icon.Move(innerOffset.Add(iconOffset))
+	hasIcon := b.icon != nil
+	hasLabel := b.label.Text != ""
+	if !hasIcon && !hasLabel {
+		// Nothing to layout
+		return
+	}
+	iconSize := fyne.NewSize(theme.IconInlineSize(), theme.IconInlineSize())
+	labelSize := b.label.MinSize()
+	padding := b.padding()
+	innerSize := size.Subtract(padding)
+	innerOffset := fyne.NewPos(padding.Width/2, padding.Height/2)
+	var iconPos fyne.Position
+	var labelPos fyne.Position
+	if hasLabel {
```

Comment:
I can't help but feel that this code could be simplified if an intermediate container was added.
Conceptually a bigger change, but I think it will be vastly simpler.

I.e. a content box that takes an icon and/or text in the right order... then all you have to do is align it.
Trying to find ways to avoid large indents of layout code like this

Reply:
Switched to using an HBoxLayout to layout at min size when both icon and label are present, then reused the alignedPosition to offset the items correctly. The code is definitely shorter.

The downside with this approach is that HBoxLayout forces all children to be the same height, and since the min height of the label is 21, the icon is resized to 20x21, causing the golden image tests and TestButtonRenderer_Layout_Stretch to fail. 1 pixel is not that noticeable, but if the text size is increased, this will cause the button image to get stretched in height.

Reply:
Oh bummer sorry, didn't see that coming. 
I think this may be a bug in icon - it should always maintain ratio in the space.
What if the icon's default image fill is ImageFillContain?

Reply:
Fixed now, thanks to your work on #919